xrootd/protocol/protocol.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package protocol contains the XRootD protocol specific types.
+package protocol // import "go-hep.org/x/hep/xrootd/protocol"
+
+import (
+	"encoding/binary"
+	"fmt"
+)
+
+// ResponseStatus is the status code indicating how the request completed.
+type ResponseStatus uint16
+
+const (
+	// Ok indicates that request fully completed and no addition responses will be forthcoming.
+	Ok ResponseStatus = 0
+	// OkSoFar indicates that server provides partial response and client should be prepared
+	// to receive additional responses on same stream.
+	OkSoFar ResponseStatus = 4000
+	// Error indicates that an error occurred during request handling.
+	// Error code and error message are sent as part of response (see xrootd protocol specification v3.1.0, p. 27).
+	Error ResponseStatus = 4003
+)
+
+// ServerError is the error returned by the XRootD server as part of response to the request.
+type ServerError struct {
+	Code    int32
+	Message string
+}
+
+func (err ServerError) Error() string {
+	return fmt.Sprintf("xrootd: error %d: %s", err.Code, err.Message)
+}
+
+// StreamID is the binary identifier associated with a request stream.
+type StreamID [2]byte
+
+// ResponseHeaderLength is the length of the ResponseHeader in bytes.
+const ResponseHeaderLength = 2 + 2 + 4
+
+// ResponseHeader is the header that precedes all responses (see xrootd protocol specification).
+type ResponseHeader struct {
+	StreamID   StreamID
+	Status     ResponseStatus
+	DataLength int32
+}
+
+// Error returns an error received from the server or nil if request hasn't failed.
+func (hdr ResponseHeader) Error(data []byte) error {
+	if hdr.Status == Error {
+		code := int32(binary.BigEndian.Uint32(data[0:4]))
```

Comment:
are we sure the input `data` slice will always be of size at least 4?
perhaps better to add a check for that?

what do you think?